facet-json/src/deserialize.rs

```diff
@@ -402,10 +402,60 @@ pub fn from_slice_wip<'input: 'facet, 'facet>(
                                         trace!("Array starting for list ({})!", wip.shape().blue());
                                         reflect!(put_default());
                                     }
+                                    Def::Struct(s) => {
+                                        if s.kind == StructKind::Tuple {
+                                            trace!(
+                                                "Array starting for tuple ({})!",
+                                                wip.shape().blue()
+                                            );
+                                            // Special handling for unit type ()
+                                            if s.fields.is_empty() {
+                                                // Check if the array is empty by peeking at the next token
+                                                let next_token = read_token!();
+                                                if next_token.node == Token::RBracket {
+                                                    // Empty array means unit type () - we're good
+                                                    reflect!(put_default());
+                                                } else {
+                                                    // Non-empty array is not valid for unit type
+                                                    bail!(JsonErrorKind::UnsupportedType {
+                                                        got: wip.innermost_shape(),
+                                                        wanted: "empty array",
+                                                    });
+                                                }
+                                            } else {
+                                                reflect!(put_default());
+                                            }
+                                        } else {
```

Comment:
In general, I would love to get to see more early returns in there to avoid such deep indentation.

Reply:
I returned to address this but see it's been modified (refactored to remove macros: 6e1d479) which I think resolved this complaint. Specifically [here](https://github.com/lmmx/facet/blob/22af2b63dd8aa63f1b1ff750f599732503087a1a/facet-json/src/deserialize.rs#L422C1-L430C38) there is now a `?` which will early return on error :smiley_cat: :+1: